### PR TITLE
fix(web): maps touch-layout chiral alt, ctrl to non-chiral when non-chiral keyboard is active

### LINKED ISSUE #6792: bug(web): right alt keys don't work on iOS and Android when referred to
**Describe the bug**

Try and install the keyboard below and see how the right alt keys, i.e. ឞ on the default layer and ឝ on the shift layer, don't output anything.

Keyboard project folder with their built package: 
[sil_jarai.zip](https://github.com/keymanapp/keyman/files/8924387/sil_jarai.zip)

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #6792.

Up until now, KeymanWeb hasn't bothered comparing a keyboard's chirality with the corresponding OSK's key "layer" property - it's simply assumed consistency between the two.  Should a keyboard developer specify a chiral "layer" property when the keyboard itself isn't actually chiral, that unchecked assumption would cause the key not to match in the expected way.  For the price of a little extra overhead, we can remove that assumption.

Note that there is no matching keyboard for the user test; instead, I've added numerous unit tests to handle the important part of the changes for the original issue.  However, it's still best to test other keyboards to make sure that others aren't affected.

### Tests

- TEST_NORMAL_CHIRALITY:  Using the "Chirality testing/bootstrapping" test page...
    - Test by using Chrome's remote-device emulation.
    - Use a longpress-shift and select the "LAlt" layer, then type at least 3 keys with the OSK.
    - Resulting text should perfectly match the keycaps for the keys you typed.
    - Then, tap shift and then longpress-shift again to select the "Shift + RAlt" layer, then type at least 3 keys with the OSK.
    - Resulting text should, again, perfectly match the keycaps for the keys you typed.

- TEST_CHIRALITY_HARDWARE:  Using the "Chirality testing/bootstrapping" test page...
    - Without remote-device emulation in Chrome, but with a hardware keyboard that has two Alt keys...
    - Hold LAlt.  The desktop OSK should change layer and show new keycaps.
    - Type at least three keys while holding LAlt.  The resulting text should perfectly match the keycaps for the keys you typed.
    - Then, hold Shift + RAlt.  The desktop OSK should change layer and show new keycaps. 
    - Again, type at least 3 keys with the OSK.  The resulting text should perfectly match the keycaps for the keys you typed.
    - Now, hold Shift + LAlt + LCtrl.  The desktop OSK should show the same keycaps that it did for Shift + RAlt.

- TEST_NORMAL_KEYBOARD_OSK:  Using the "Test unminified Keymanweb" test page...
    - Test by using Chrome's remote-device emulation.
    - Select the "French" keyboard.
    - Type at least two keys on the base layer and confirm they act as expected.
    - Then, use a longpress-shift to select the Alt+Ctrl layer, then type `[` and `]`, confirming they act as expected.
    - Then, type the `~` key.  Nothing should emit.
    - Then, drop to the base layer, then type an `a`.  An `ã` should result.

- TEST_NORMAL_KEYBOARD_HARDWARE:  Using the "Test unminified Keymanweb" test page...
    - Without remote-device emulation in Chrome, but with a hardware keyboard that has two Alt keys...
    - Select the "French" keyboard.
    - Type at least two unmodified (no Shift, Ctrl, Alt...) _with your hardware keyboard_ and confirm they act as expected.
    - Then, hold Alt + Ctrl and type your hardware `5` and `-` keys - they should produce `[` and `]` respectively, regardless of which Alt and Ctrl key you use.
    - Then, type Alt + Ctrl + `2` key.  Nothing should emit.
    - Then, release Alt + Ctrl, then type an `a`.  An `ã` should result.

### CODE REVIEW

File: common/web/keyboard-processor/src/text/kbdInterface.ts [hunk 1]
```diff
@@ -529,6 +529,43 @@ namespace com.keyman.text {
...
 
+    /**
+     * Maps a KeyEvent's modifiers to their appropriate value for key-rule evaluation
+     * based on the rule's specified target modifier set.
+     *
+     * Mostly used to correct chiral OSK-keys targeting non-chiral rules.
+     * @param e The source KeyEvent
+     * @returns
+     */
+    private static getEventRuleModifiers(e: KeyEvent, targetModifierMask: number): number {
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/text/kbdInterface.ts:540]: Given we only ever use `e.Lmodifiers` in the `KeyEvent`, why don't we make this function more "generic" in nature and pass in that value? Rename it to `transformModifierChiralityModeForTarget` or something like that?

```suggestion
    private static getEventRuleModifiers(eventModifiers: number, targetModifierMask: number): number {
```

File: common/web/keyboard-processor/src/text/kbdInterface.ts [hunk 2]
```diff
@@ -529,6 +529,43 @@ namespace com.keyman.text {
...
+      let modifiers = e.Lmodifiers;
+
+      // If the target rule does not use chiral alt...
+      if(!(targetModifierMask & CHIRAL_ALT)) {
+        const alt_intersection  = modifiers & CHIRAL_ALT;
+
+        if(alt_intersection) {
+          // Undo the chiral part          and replace with non-chiral.
+          modifiers ^= alt_intersection  | Codes.modifierCodes["ALT"];
+        }
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/text/kbdInterface.ts:553]:
```suggestion
        const altIntersection  = modifiers & CHIRAL_ALT;

        if(altIntersection) {
          // Undo the chiral part         and replace with non-chiral.
          modifiers ^= altIntersection  | Codes.modifierCodes["ALT"];
        }
```

Should be using `camelCase` :grin:

File: common/web/keyboard-processor/src/text/kbdInterface.ts [hunk 3]
```diff
@@ -529,6 +529,43 @@ namespace com.keyman.text {
...
+     *
+     * Mostly used to correct chiral OSK-keys targeting non-chiral rules.
+     * @param e The source KeyEvent
+     * @returns
+     */
+    private static getEventRuleModifiers(e: KeyEvent, targetModifierMask: number): number {
+      const CHIRAL_ALT  = Codes.modifierCodes["LALT"]  | Codes.modifierCodes["RALT"];
+      const CHIRAL_CTRL = Codes.modifierCodes["LCTRL"] | Codes.modifierCodes["RCTRL"];
+
+      let modifiers = e.Lmodifiers;
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/text/kbdInterface.ts:544]:
```suggestion
      let modifiers = eventModifiers;
```

If you go with my earlier suggestion... (Or, remove the `modifiers` variable altogether and just work with the `eventModifiers` input parameter, which is probably better.)

File: common/web/keyboard-processor/src/text/kbdInterface.ts [hunk 4]
```diff
@@ -529,6 +529,43 @@ namespace com.keyman.text {
...
+      }
+
+      // If the target rule does not use chiral ctrl...
+      if(!(targetModifierMask & CHIRAL_CTRL)) {
+        const ctrl_intersection = modifiers & CHIRAL_CTRL;
+
+        if(ctrl_intersection) {
+          // Undo the chiral part          and replace with non-chiral.
+          modifiers ^= ctrl_intersection | Codes.modifierCodes["CTRL"];
+        }
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/text/kbdInterface.ts:563]:
```suggestion
        const ctrlIntersection = modifiers & CHIRAL_CTRL;

        if(ctrlIntersection) {
          // Undo the chiral part         and replace with non-chiral.
          modifiers ^= ctrlIntersection | Codes.modifierCodes["CTRL"];
        }
```

**Review by mcdurdin — COMMENTED** [review]
This looks good. I was slightly concerned about additional performance impact of another function call for every rule match but I don't think it'll be measurable.

Only request is to make the function more generic in nature given it really doesn't need a `KeyEvent`. (That also simplifies the tests slightly -- can eliminate `key` and `BASE_KEY_EVENT` variables)

Could we also request a user test with the sil_jarai keyboard from the original issue? We can add it to the repo in common/test/keyboards, and it should be built automatically as an installable artifact for iOS/Android then.

**Comment by jahorton** [comment]
> Could we also request a user test with the sil_jarai keyboard from the original issue?  We can add it to the repo in common/test/keyboards, and it should be built automatically as an installable artifact for iOS/Android then.

The thing is... what I received with the .zip had package compilation errors, at least with Developer 15.0.260.  The _keyboard_ was fine, but the package (so, for iOS/Android) was not.

I'm also worried about _exactly_ how to document / annotate it, since we need this specific in-development version of `sil_jarai`, not anything updated after the discussion on the original issue.

File: common/web/keyboard-processor/src/text/kbdInterface.ts [hunk 5]
```diff
@@ -547,13 +584,15 @@ namespace com.keyman.text {
       var modifierBitmask = bitmask & Codes.modifierBitmasks["ALL"];
       var stateBitmask = bitmask & Codes.stateBitmasks["ALL"];
 
+      const eventModifiers = KeyboardInterface.getEventRuleModifiers(e, Lruleshift);
```
Comment (jahorton) [inline common/web/keyboard-processor/src/text/kbdInterface.ts:587]:
```suggestion
      const eventModifiers = KeyboardInterface.getEventRuleModifiers(e.Lmodifiers, Lruleshift);
```

**Comment by jahorton** [comment]
> This looks good. I was slightly concerned about additional performance impact of another function call for every rule match but I don't think it'll be measurable.

The `fullContextMatch` method (`.KFCM` for short) we added in KMW 10 is easily far worse in that direction, given all its branching paths.  (Granted, not all rules need it.)  In comparison, this function should be _very_ consistent in what branches need to be taken, especially for non-chiral keyboards.  Even for chiral keyboards, I'd expect only two missed branch predictions per call, tops, unless they made the same mistake that led to #6792.

If we _really_ care, there should even be a way to add optimization in Developer's web compiler to help reduce that!  Note how the top-level `if` statements all condition on a logical-and with the rule's modifier bitmask.  So, if we group rules by their expected modifier masks, that'll dramatically cut down on branch prediction misses.  Though... that's _probably_ overkill at the moment.

**Comment by mcdurdin** [comment]
> The thing is... what I received with the .zip had package compilation errors, at least with Developer 15.0.260. The _keyboard_ was fine, but the package (so, for iOS/Android) was not.

So we'll need to fix the errors I guess. Shouldn't be too hard.

> I'm also worried about _exactly_ how to document / annotate it, since we need this specific in-development version of `sil_jarai`, not anything updated after the discussion on the original issue.

We include the version that was in the zip file in the common/test/keyboards folder in this repo. I think that resolves that issue. :-)

> The `fullContextMatch` method (`.KFCM` for short) we added in KMW 10 is easily far worse in that direction, given all its branching paths.

Oh for sure. And we can see perf as an issue for some complex keyboards already. So that's why I am hoping not to make it worse.

And for sure we can optimise the compiler better. But we need to be really careful with any algorithmic optimisations -- subtle changes in behaviour do come back to bite us (e.g. the compiler change which grouped rules with identical keys actually caused trouble where there were chirality mismatches in the same keyboard -- rules were then missed or compared out of order. In a simple example, y+a would never match in KeymanWeb because of the grouping of the two ralt rules:
```
+ [RALT K_A] > 'foo'
'x' + [RALT K_A] > 'bar'
'y' + [ALT K_A] > 'baz'
```
)

**Comment by jahorton** [comment]
> I would like us to make sure we test all three pathways: touch, desktop OSK, and hardware, given that this is a modification to the core logic of KMW. I know it's a pain... but any oversight here will have significant ramifications.

What sort of tests are you thinking in that regard?

**Comment by mcdurdin** [comment]
> What sort of tests are you thinking in that regard?

Basic functionality tests for input -- make sure that modifiers and layers are working correctly, particularly all the little edge cases, e.g. layer overrides, caps support, etc. I can't see any issues with the change, but we've gone through this kind of thing before and thought it was all good and only discovered much later some aspect we've forgotten.

I think we should be able to use existing keyboards for most of these tests, e.g. khmer_angkor, basic_french, etc.

**Comment by bharanidharanj** [comment]
* **TEST_NORMAL_CHIRALITY (PASSED):** Tested in Chrome's remote device - emulation and it is working as expected. 

LAlt Layer:
 

RAlt Layer:
 


* **TEST_NORMAL_KEYBOARD_OSK (PASSED):** Tested in Chrome's remote device - emulation and it is working as expected.

**Comment by jahorton** [comment]
@mcdurdin I've added two new test cases to support both nonchiral and chiral keyboards for desktop form factors.  Does this seem reasonably sufficient now, or is there another angle or two I should consider adding?

**Comment by mcdurdin** [comment]
> @mcdurdin I've added two new test cases to support both nonchiral and chiral keyboards for desktop form factors. Does this seem reasonably sufficient now, or is there another angle or two I should consider adding?

I *think* so? Only thing I am wondering is if we are covering both OSK and hardware interactions sufficiently?

**Review by mcdurdin — APPROVED** [review]
LGTM now. Let's keep an eye on this change through the 16.0 cycle as modifier comparisons are really complicated (complicated history in so many products, not just KeymanWeb).

**Comment by jahorton** [comment]
> LGTM now. Let's keep an eye on this change through the 16.0 cycle as modifier comparisons are really complicated (complicated history in so many products, not just KeymanWeb).

Wouldn't we want to 🍒-pick this for 15.0, though?  Or are you saying that we feel it's potentially a bit too risky for that?

**Comment by jahorton** [comment]
> I don't think we should cherry-pick to stable. This is a change to how we handle chirality for which we have a solid workaround. It's important because it improves consistency cross-platform, but I think it should be considered a potentially breaking change, not something that we can back-port.

Huh.  My personal feelings are that it should be safe to 🍒... but as noted earlier in the comment chain, it's something easy to prevent entirely during keyboard development, and it's not something that users generally run into.  So... I guess there's not really any urgency to 🍒 it; any currently-affected keyboards need adjustment to work well with older versions anyway.

So... yeah, I guess there's no major loss in not 🍒-ing it at this time, even if it were proven to be 100% safe.

**Comment by MakaraSok** [comment]
# Test Results

All tests done on iPad mini 6 running iOS 15.5.

* **TEST_CHIRALITY_HARDWARE (FAILED):** 
In Chrome on macOS 12.4, the keycaps shown don't correspond with keys triggered and fired on the hardware keyboard.

Type keys <kbd>e</kbd><kbd>r</kbd><kbd>t</kbd><kbd>y</kbd><kbd>u</kbd> while holding LAlt on the hardware keyboard, output `´®ʈɥ¨` instead of the expected `ɛɽʈɥʊ`. See the comparison below.

| With | Keys | actual output| screenshot of actual output | expected output |
| --- | --- | --- | --- | --- |
| OSK | LAlt + <kbd>e</kbd><kbd>r</kbd><kbd>t</kbd><kbd>y</kbd><kbd>u</kbd> | `ɛɽʈɥʊ ` |  | `ɛɽʈɥʊ` |
| Hardware | LAlt + <kbd>e</kbd><kbd>r</kbd><kbd>t</kbd><kbd>y</kbd><kbd>u</kbd>  | `´®ʈɥ¨` |  | `ɛɽʈɥʊ` |

In Chrome on Windows 10 Pro 21H2, the keycaps shown do correspond with keys triggered and fired on the hardware keyboard.

Failed this test for now as it doesn't work on macOS. 

* **TEST_NORMAL_KEYBOARD_HARDWARE (PASSED):** In Chrome on macOS 12.4 and Windows 10 Pro 21H2, each keys indicated produce the expected characters/outputs.
image

**Comment by jahorton** [comment]
Huh... we never noticed that before; I guess something's a bit funny with the keymappings for that case?

I just attempted to reproduce that FAILed test on `master`... and _succeeded_.  So, at least it isn't something caused by this PR.  So, yay, new issue I guess?  But at least it matches prior behavior.

So, given that, I believe the correct call is to say:

TEST_CHIRALITY_HARDWARE:  PASSED - though with discovery of a pre-existing bug, recorded as #6876.

**Comment by jahorton** [comment]
Well... I kind of forgot about this, as did everyone else.  Nobody chimed in with extra issues, so I suppose that means it's merging time.

**Comment by mcdurdin** [comment]
> as did everyone else

I hadn't forgotten, nor had status.keyman.com :grin:. But I was leaving it to you. We had bigger fish to fry last week for me to chase it with you 😉
